add plugin page (#463)
aiidalab/aiidalab-qe: docs/source/development/architecture.rst

.. _develop:architecture:

************************
Architecture
************************

Wizards UI
==========

> Quantum ESPRESSO app (QeApp) uses the Wizards UI, which divides one calculation into four steps. Each step may contain several sections (panels), as shown below.

Author: I prefer to use the `Wizards UI` instead of the class name `WizardAppWidget`. Here we give the overview of the architecture instead of detailed code.

.. _develop:architecture:

************************
Architecture
************************

Wizards-like UI
================

> QuantumESPRESSO app uses the [Wizards-like UI design](https://en.wikipedia.org/wiki/Wizard_(software)), which divides one calculation into four steps.

Reviewer: ```suggestion
QuantumESPRESSO app uses the `Wizards-like UI design <https://en.wikipedia.org/wiki/Wizard_(software)>`_, which divides one calculation into four steps.
```

Here we use rst not markdown.

            "initial_magnetic_moments": None,
            "pw": {
                "parameters": {
                    "SYSTEM": {"ecutwfc": 30.0, "ecutrho": 240.0, "tot_charge": 0.0}
                },
                "pseudos": {"Si": "eaef3352-2b0e-4205-b404-e6565a88aec8"},
            },
            "pseudo_family": "SSSP/1.2/PBEsol/efficiency",
            "kpoints_distance": 0.5,
        },
        "bands": {"kpath_2d": "hexagonal"},
        "pdos": {...},
        "plugin_1": {...},
        "plugin_2": {...},
    }

Plugin
=========
QuantumESPRESSO app supports running multiple properties (bands, pdos, etc.) calculations in one app.
The individual properties to be developed and seamlessly integrated into the QuantumESPRESSO app as plugins. This integration is made possible due to several key aspects:

- the configuration for a property calculation has its settings unrelated to other properties.
- the sub-workchain of the properties can be run independently.
- the analysis of the results of the properties is independent.

Each plugin responsible for one property calculation.
For instance, we could create a PDOS plugin, including its settings, workchain, and result analysis.

> The GUI of the PDOS plugin is only loaded when the user selects to run it.

Reviewer: ```suggestion
The GUI of the PDOS plugin is only loaded when the user selects to run it, by ticking the checkbox in the workflow setting.
```

Reviewer: Most are indentation fixes and a hyperlink fix. Please see if you think this is better phrased. In case you don't know, there is a tip in github that you can use `Commit suggestion` button above to quickly accept and submit the commit.

Author: I'd rather not add this `, by ticking the checkbox in the workflow setting.` . This is the introduction of the architecture, it's better not to go into the detail usage. 
I know the `commit suggestion`, thanks.

Reviewer: Okay~

.. _develop:architecture:

************************
Architecture
************************

Wizards UI
==========

QuantumESPRESSO app uses the Wizards UI, which divides one calculation into four steps.
Each step may contain several sections (panels), as shown below.

.. image:: ../_static/images/plugin_step.png

Parameter transfer between steps
---------------------------------

The data is passed to the next step by linking it to the corresponding ``trail`` of the step.
For example, the ``confirmed_structure`` of step 1 is linked to the ``input_structure`` ``trail`` of step 2.

.. code:: python

    ipw.dlink(
            (self.structure_selection_step, "confirmed_structure"),
            (self.configure_qe_app_work_chain_step, "input_structure"),
        )

> In the configuration step, there are several panels.

Author: :+1:

> The parameters from these panels are generated and stored as a dictionary,
> which is linked to the ``input_parameters`` ``trail`` of the next submit step.

Reviewer: ```suggestion
The parameters from these panels are generated and stored as a dictionary, which is linked to the ``input_parameters`` trait of the next submit step widget.
```

The dictionary has the following structure:

.. code:: python

    {
        "workchain": {
            "protocol": "fast",
            "relax_type": "positions",
            "properties": ["bands", "pdos", "relax"],
            "spin_type": "none",
            "electronic_type": "insulator",
        },
        "advanced": {
            "initial_magnetic_moments": None,
            "pw": {
                "parameters": {
                    "SYSTEM": {"ecutwfc": 30.0, "ecutrho": 240.0, "tot_charge": 0.0}
                },
                "pseudos": {"Si": "eaef3352-2b0e-4205-b404-e6565a88aec8"},
            },
            "pseudo_family": "SSSP/1.2/PBEsol/efficiency",
            "kpoints_distance": 0.5,
        },
        "bands": {"kpath_2d": "hexagonal"},
        "pdos": {...},
        "plugin_1": {...},
        "plugin_2": {...},
    }

Plugin
======

QuantumESPRESSO app supports running multiple properties (bands, pdos, etc.) calculations in one app.

> Please take into account the following facts:

Reviewer: Is this a rule for developers or it is a thing to note?

Author: I changed it to ```This integration is made possible due to several key aspects:```

- the configuration for a property calculation has its settings unrelated to other properties.
- the sub-workchain of the properties can be run independently.

> - the analysis of the results of the properties is also independent.

Reviewer: ```suggestion
- the analysis of the results of the properties is independent.
```

Thus, we can develop a property separately and integrate it into the QuantumESPRESSO app as a plugin.
Each plugin responsible for one property calculation.
For example, we could create a PDOS plugin, including its settings, workchain, and result analysis.

> The GUI of the PDOS plugin is only loaded when the user selects to run the PDOS property.

Reviewer: ```suggestion
The GUI of the PDOS plugin is only loaded when the user selects it.

Here is an example, where two new setting panels are shown when the user selects to run the properties.

> .. figure:: ../_static/images/plugin_example.png

Reviewer: See my comment on this file, I think it maybe better to have two static figures to show before and after the property being selected?

> A QuantumESPRESSO app plugin will typically register new panels (setting, result), and workchain to extend the app's functionality.

Reviewer: ```suggestion
A QuantumESPRESSO app plugin will register new corresponding panels (setting, result), and workchain to extend functionalities of the App.
```

Author: I prefer to keep `typically`, and not add `corresponding`.

The plugin design makes the QuantumESPRESSO app more modularized and pluggable.

> So the developer can maintain their plugin as a separate folder in the QuantumESPRESSO app (even a separate package).

Reviewer: ```suggestion
As a result, the developer can maintain their plugin in a separate folder in the QuantumESPRESSO app code base or even as a separate package.
```
